### DISCUSSION TITLE
How to compute simple Biot-Savart from segment for maximum speed

### DISCUSSION THREAD

**Discussion by piramiday** [magpylib/magpylib#591] [Q&A] [answered]
I am computing the Biot-Savart B field from a single segment on a lot of points, say, 30M.
My custom routine for Biot-Savart integration appears quite faster than the otherwise-excellent `magpylib`, and I was wondering whether I am doing something wrong.

I am instantiating `magpylib.current.Line` to compute the B field from a long (virtually infinite) filament along the z axis, like so:
```python
import numpy as np
import magpylib

xyz = np.zeros((2, 3))
xyz[:, 2] = np.linspace(-100, 100, 2)

src = magpylib.current.Line(current=1, vertices=xyz)
```
and then I am timing the call to get:
```
field = src.getB(obs)
```
Am I doing something inefficient?
I might have better performance chunking off the big `obs` vector, but then again, that is not what my library is currently doing.

The only note worth mentioning is that I had spent some time tweaking my Biot-Savart library to remove, for example, any call to `np.linalg.norm` that was horribly slow, together with replacing `np.sqrt` with a more direct `** 2` operation, and so on.

If I try and go towards using the `core` package, it seems to me that I have to repeat my input current and input coordinates for 30M times, which feels extremely inefficient.
Is that what the `getB` call is doing?

Any suggestion is appreciated.

**Comment by OrtnerMichael** [comment, marked as answer]
hi @piramiday

sorry for the late answer - this was a terrible week at work :).

So, there are a lot of reasons why Magpylib might underperform:
1. When using the object oriented interface it takes all inputs, tiles them up, computes relative position and orientations, then forwards to the computation core, computes, then again does everything backwards. If you use the core instead, you will save a lot of time. The reason is that the object oriented interface is very convenient to use, but this comes at a price.
2. As you pointed out our core routine ```current_line_field()``` is not so well implemented ... a lot of expensive calls like `np.all()`, `np.any()`, `np.linalg.norm()` and multiple masks- those are slowing things down for sure. A lot can be improved here.
3. The reason for those slow `all` and `any` calls and the masking thats going on are numerical instabilities of the simple expression when computing with Biot-Savart. There are a lot of special cases where the analytic expression fails, returns nan or infty, e.g. when the observer lies on the line, or when the line segment has zero length, or when the observer is at large distance, or when approaching the singularities. To make sure that the special cases don't ruin the computation we have to apply those checks - and this slows things down :(

Many special cases we don't even treat at this point. Look at the following plot generated by this code:

```
import matplotlib.pyplot as plt
import numpy as np
import magpylib as magpy

line = magpy.current.Line(1, ((-1,0,0), (0,0,1)))
xs = np.logspace(-20, 20, 10000)
obs = [(x,0,1) for x in xs]
B = line.getB(obs)
plt.plot(xs, abs(B))
plt.gca().set(
    xscale='log',
    yscale='log',
)
plt.show()
```

![image](https://user-images.githubusercontent.com/47384012/221413919-f6a988c1-e830-47db-905a-ff0ddf672bbc.png)

on the left you see the numerical breakdown when approaching the singularity, on the right you see the breakdown at large distances.

so long story short ... there are quite some issues with the analytic expressions. Any suggestions how to improve are very welcome :). Max your performance by using the core instead of the object oriented interface. In any case the inputs must be tiled - no way around this.
